### PR TITLE
Mono 2017-02 integration

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

**Comment by marek-safar** [comment]
@spouliot what's the plan how to get this merged?

File: mk/xamarin.mk [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ifdef ENABLE_XAMARIN
-NEEDED_MACCORE_VERSION := cebe977dc69e3ed628f3c87e349d971f52dd4b34
-NEEDED_MACCORE_BRANCH := master
+NEEDED_MACCORE_VERSION := 874818b0df27469298fcdb4b5fed61346c57f8b9
+NEEDED_MACCORE_BRANCH := mono-2017-02
```
Comment (rolfbjarne) [inline mk/xamarin.mk]: Any maccore changes should be merged to maccore/master first (assuming xamarin-macios/master works with those changes).

**Comment by spouliot** [comment]
@marek-safar beside reviews QA has started testing the branch (not identical) on devices last week. Right now there's only one bug I'm aware of:

* https://bugzilla.xamarin.com/show_bug.cgi?id=53293

We still have to run the networking tests to ensure there's no regression. If/when all is green then it meets _my releasable_ threshold. Some external thumbs up from @kumpera and @jonpryor will also be needed (to make sure we all plan to release the same mono for 15.2).

**Comment by spouliot** [comment]
@jstedfast, @VincentDondain can you please double check the changes to the msbuild files. If needed also ping the XVS team to make them aware of the changes. Thanks!

**Comment by spouliot** [comment]
@chamons there are XM specific changes to review and I also want your 👍 for XM :)

File: mk/xamarin.mk (hunk 1 above)
Comment (marek-safar) [inline mk/xamarin.mk]: Pending PR https://github.com/xamarin/maccore/pull/556

**Comment by jstedfast** [comment]
I take it that these patches drop mcs in favor of csc?

The msbuild patches look good to me.

**Comment by rolfbjarne** [comment]
@spouliot it looks like you commented on the PR itself (as opposed to specific lines), which makes it hard to determine what you're comments are referring to.

**Comment by spouliot** [comment]
@rolfbjarne weird, I was reviewing commits-by-commits, not the single-PR changes. Never noticed it placed them in the PR, not the source.

**Comment by rolfbjarne** [comment]
There are test build failures:

> 	/Users/builder/jenkins/workspace/xamarin-macios-pr-builder/external/mono/mcs/class/System/Test/System.Diagnostics/ProcessTest.cs(1165,9): error CS0103: The name `RunningOnUnix' does not exist in the current context

**Comment by rolfbjarne** [comment]
The `MonoTouch.PInvokeCallbackAttribute` class is gone from System.dll:

```
$ monodis --typedef /Library/Frameworks/Xamarin.iOS.framework/Versions/10.5.0.479/lib/mono/Xamarin.iOS/System.dll | grep MonoTouch.MonoPInvokeCallback
$ monodis --typedef /Library/Frameworks/Xamarin.iOS.framework/Versions/9.8.0.156/lib/mono/Xamarin.iOS/System.dll | grep MonoTouch.MonoPInvokeCallback
796: MonoTouch.MonoPInvokeCallbackAttribute (flist=5437, mlist=6757, flags=0x100000, extends=0x4d)
```

This is a breaking change.

**Comment by marek-safar** [comment]
@rolfbjarne how is MonoTouch.PInvokeCallbackAttribute change breaking when the type was not visible outside of System assembly?

**Comment by rolfbjarne** [comment]
`using MonoTouch;` fails to compile now, because `MonoTouch.PInvokeCallbackAttribute` was the only type in the `MonoTouch` namespace.

**Comment by marek-safar** [comment]
@rolfbjarne Fixed both issue
@spouliot Fixed https://bugzilla.xamarin.com/show_bug.cgi?id=53293, not sure what are the other commends about

**Comment by rolfbjarne** [comment]
@marek-safar Looking at the api diff (https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/3300/API_diff/), it looks like we're adding all the System.Reflection.Emit types. Is that intentional?

**Comment by marek-safar** [comment]
@rolfbjarne These types were added in Mono 4.8.

**Comment by rolfbjarne** [comment]
@kzu This PR will change our debug files from .dll.mdb to .pdb, will that affect VS support in any way?

**Comment by spouliot** [comment]
@kzu @josegallardo @emaf beside @rolfbjarne question above there are also other msbuild related changes in this PR that we'll like you to review before merging

**Comment by spouliot** [comment]
@marek-safar this can't be merged (conflicts) in it's current state. Can you merge master into `mono-2017-02` so we can have a (hopefully) last build and test results.

**Comment by spouliot** [comment]
https://bugzilla.xamarin.com/show_bug.cgi?id=53481 is presently blocking QA to complete testing

**Comment by spouliot** [comment]
@kzu @josegallardo @emaf I have not heard any comment / go ahead on the msbuild parts. A quick thumb up would be appreciated. thanks!

**Comment by rolfbjarne** [comment]
build

**Comment by kzu** [comment]
@rolfbjarne @spouliot the PR looks fine. We'll just need to delete the PDB2MDB stuff on our side. I assume this will also be the case for XA, @jonpryor? 

This basically means we'll likely be able to remove altogether the windows-only targets that XVS installs, I hope. 

We'll adapt our side once this is merged.

Thanks and sorry for the delay!

**Pull request merged**